core/new-gui/src/app/workspace/service/workflow-result/workflow-result.service.ts

```diff
@@ -0,0 +1,165 @@
+    this.wsService.subscribeToEvent('WebResultUpdateEvent').subscribe(event => this.handleResultUpdate(event.updates));
+  }
+
+  public getResultUpdateStream(): Observable<Record<string, WebResultUpdate>> {
+    return this.resultUpdateStream;
+  }
+
+  public getPaginatedResultService(operatorID: string): OperatorPaginationResultService | undefined {
+    return this.paginatedResultServices.get(operatorID);
+  }
+
+  public getResultService(operatorID: string): OperatorResultService | undefined {
+    return this.operatorResultServices.get(operatorID);
+  }
+
+  private handleResultUpdate(event: WorkflowResultUpdate): void {
+    Object.keys(event).forEach(operatorID => {
+      const update = event[operatorID];
+      if (isWebPaginationUpdate(update)) {
+        const paginatedResultSerivce = this.getOrInitPaginatedResultService(operatorID);
+        paginatedResultSerivce.handleResultUpdate(update);
+      } else if (isWebDataUpdate(update)) {
+        const resultSerivce = this.getOrInitResultService(operatorID);
+        resultSerivce.handleResultUpdate(update);
+      } else {
+        const _exhasutiveCheck: never = update;
+      }
+    });
+    this.resultUpdateStream.next(event);
+  }
+
+  private getOrInitPaginatedResultService(operatorID: string): OperatorPaginationResultService {
+    let service = this.getPaginatedResultService(operatorID);
+    if (!service) {
+      service = new OperatorPaginationResultService(operatorID, this.wsService);
+      this.paginatedResultServices.set(operatorID, service);
+    }
+    return service;
+  }
+
+  private getOrInitResultService(operatorID: string): OperatorResultService {
+    let service = this.getResultService(operatorID);
+    if (!service) {
+      service = new OperatorResultService(operatorID);
+      this.operatorResultServices.set(operatorID, service);
+    }
+    return service;
+  }
+}
+
+export class OperatorResultService {
+
+  private chartType: ChartType | undefined;
+  private resultSnapshot: ReadonlyArray<object> = [];
+
+  constructor(public operatorID: string) {
+  }
+
+  public getCurrentResultSnapshot(): ReadonlyArray<object> {
+    return this.resultSnapshot;
+  }
+
+  public getChartType(): ChartType | undefined {
+    return this.chartType;
+  }
+
+  public handleResultUpdate(update: WebDataUpdate): void {
+    this.chartType = update.chartType;
+    if (update.mode.type === 'SetSnapshotMode') {
+      this.resultSnapshot = update.table;
+    } else if (update.mode.type === 'SetDeltaMode') {
+      const combinedResult = this.resultSnapshot.concat(update.table);
+      this.resultSnapshot = combinedResult;
+    } else {
+      const _exhaustiveCheck: never = update.mode;
+    }
+  }
+}
+
+class OperatorPaginationResultService {
+
+  private pendingRequests: Map<string, Subject<PaginatedResultEvent>> = new Map();
+  private resultCache: Map<number, ReadonlyArray<object>> = new Map();
+  private currentPageIndex: number = 1;
```

Comment:
still suggest changing from `pageIndex` to `pageNum` to be more intuitive.

Reply:
I checked the library and the name `pageIndex` comes from there. Therefore we'd better be consistent with the library

Reply:
ok